tests/lib/fortigate_test.py

```diff
@@ -0,0 +1,390 @@
+# Copyright 2019 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unittest for fortigate policy rendering module."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import string
+import re
+import unittest
+
+from capirca.lib import fortigate
+from capirca.lib import nacaddr
+from capirca.lib import naming
+from capirca.lib import policy
+import mock
+
+
+GOOD_HEADER = """
+header {
+  comment:: "this is a test acl"  
+  target:: fortigate from-id 2
+}
+"""
+
+BAD_HEADER = """
+header {
+  comment:: "this is a test acl"
+  target:: fortigate edge-filter
+}
+"""
+
+TERM_TEMPLATE = """
+term good-term-2 {{
+  source-interface:: {src_interface}
+  destination-interface:: {dest_interface}
+  protocol:: {protocol}
+  destination-address:: {dest_addr}
+  destination-port:: {dest_port}
+  source-address:: {src_addr}
+  source-port:: {src_port}
+  action:: {action}
+  logging:: {logging}
+}}
+"""
+
+
+class CustomFormatter(string.Formatter):
+  DEFAULT_VALUES = {
+      'src_interface': 'wan1',
+      'dest_interface': 'wan2',
+      'protocol': 'tcp',
+      'src_addr': 'SOME_HOST',
+      'dest_addr': 'SOME_HOST',
+      'src_port': 'HTTP',
+      'dest_port': 'HTTP',
+      'action': 'accept',
+      'logging': 'true'
+  }
+
+  def format(*args, **kwargs):
```

Comment:
This should have a bounded self argument right?

Reply:
this is an overriding of `string.Formatter.format` so i did not modified the function parameters.

Reply:
https://docs.python.org/3.6/library/string.html#string.Formatter.format
Seems it takes a bound argument to me. What docs are you reading?

Reply:
This was resolved without comment. In 3.6 self was defined within the method itself. In 3.8 it is in the method signature.
https://github.com/python/cpython/blob/3.8/Lib/string.py#L162

I also find this all kind of weird. You inherit from a class but then invoke the class method by its class identifier. Something like the following would be simpler here

```
  def Format(self, string, **kwargs):
    if 'remove_fields' in kwargs:
      for field in kwargs['remove_fields']:
        remove_regex = '.*' + field + '.*'
        string = re.sub(remove_regex, '', string)
    return self.format(string, **kwargs)
```